uicomponent-compose/timetable2021/src/main/java/io/github/droidkaigi/feeder/timetable2021/TimetableDetailScreen.kt

```diff
@@ -0,0 +1,623 @@
+                Text(text = "MOVIE")
+            }
+            Spacer(modifier = Modifier.width(16.dp))
+        }
+        if (showSlidesButton) {
+            Button(
+                modifier = Modifier.width(144.dp),
+                colors = buttonColors(
+                    backgroundColor = MaterialTheme.colors.secondary,
+                    contentColor = MaterialTheme.colors.onSecondary,
+                ),
+                onClick = {
+                    onOpenUrl(asset.slideUrl!!.toUri())
+                }
+            ) {
+                Icon(Icons.Outlined.PhotoLibrary, contentDescription = "Open Slides")
+                Spacer(modifier = Modifier.width(8.dp))
+                Text(text = "SLIDES")
+            }
+        }
+    }
+    Spacer(modifier = modifier.height(24.dp))
+}
+
+@Composable
+private fun TimetableDetailSpeakers(
+    modifier: Modifier = Modifier,
+    speakers: List<TimetableSpeaker>,
+) {
+    Text(
+        modifier = modifier,
+        text = "Speaker",
+        style = MaterialTheme.typography.subtitle2,
+    )
+    Spacer(modifier = modifier.padding(16.dp))
+
+    speakers.forEach { speaker ->
+        if (speaker.iconUrl.isNotEmpty()) {
+            Surface(
+                modifier = modifier.size(60.dp),
+                shape = CircleShape
+            ) {
+                NetworkImage(
+                    url = speaker.iconUrl,
+                    modifier = Modifier.fillMaxSize(),
+                    contentScale = ContentScale.Fit,
+                    contentDescription = "Contributor Icon"
+                )
+            }
+            Spacer(
+                modifier = modifier.padding(16.dp),
+            )
+        }
+
+        Text(
+            modifier = modifier,
+            text = speaker.name,
+            style = MaterialTheme.typography.subtitle1,
+        )
+        Text(
+            modifier = modifier,
+            text = speaker.tagLine,
+            style = MaterialTheme.typography.subtitle2,
+        )
+        Spacer(
+            modifier = modifier.padding(vertical = 16.dp),
+        )
+        Text(
+            modifier = modifier,
+            text = speaker.bio,
+            style = MaterialTheme.typography.body1,
+        )
+        Spacer(
+            modifier = modifier.padding(vertical = 16.dp),
+        )
+    }
+}
+
+private fun createSessionDate(
+    startsAt: Instant,
+    endsAt: Instant,
+): String {
+    val start = startsAt.toLocalDateTime(currentSystemDefault()).toJavaLocalDateTime()
+    val end = endsAt.toLocalDateTime(currentSystemDefault()).toJavaLocalDateTime()
+    val day = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM).format(start)
+    val startTime = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT).format(start)
+    val endTime = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT).format(end)
+
+    return "$day $startTime-$endTime"
+}
+
+private fun createCategoryIcon(
```

Comment:
It was annoying, but I am getting the icon png from the category id in the UI layer.

Reply:
For these functions, the idea is to create a state class, pass it to the Composable function, and have the following functions in it 👀 
What do you think??

```kotlin
@Composable
private fun TimetableDetailSessionInfo(
    modifier: Modifier = Modifier,
    title: String,
    dateTime: String,
    language: String,
    category: TimetableCategory,
)
```

↓

```kotlin
class TimetableDetailSessionInfoState(
    title: String,
    dateTime: String,
    language: String,
    category: TimetableCategory,
) {
  val categoryIconId: Int by lagy {
...
  }
}


@Composable
private fun TimetableDetailSessionInfo(
    modifier: Modifier = Modifier,
    state: TimetableDetailSessionInfoState
)
```

Reply:
Just come here from Twitter :D. I didn't read the full code, but how about having `categoryIconResId` as extension value of `TimetableCategory`:

```kotlin
class TimetableCategory {
  ... /* class body */
}

// Extension values
@DrawableRes
val TimetableCategory.iconResId: Int get() = ...
```

Reply:
Thanks for the comment 🙏 
Indeed, one category corresponds to one category icon id, which can be used elsewhere. So in this case, an extension is a good idea. 👍 

Apart from that, I would like to see a mechanism that can be separated from the logic.
For example, in the FeedScreen implementation, a large number of arguments are passed and implemented inside. Ideally, I think it would be better to work on modeling, but I can't imagine how that solution would work in everyday development. 😅 
For this reason, I thought it might be a good idea to provide such a state class to force myself to fix it a bit.
https://github.com/DroidKaigi/conference-app-2021/blob/main/uicomponent-compose/feed/src/main/java/io/github/droidkaigi/feeder/feed/FeedScreen.kt#L176

Reply:
@takahirom 
Indeed, TimetableDetailSessionInfo has too many arguments.
I thought about splitting the function or not making it a function, but I didn't think it was a good idea.

I'll try to put it in a private data class as you suggested.
Thanks for your suggest!

---

@eneim 
Thank you for your comment.
I will try to make it a private extended property once.